services/socks5/service/service.go

```diff
@@ -0,0 +1,252 @@
+	if m.servicePort == 0 {
+		return nil, errors.New("Service port not initialized")
+	}
+
+	sessionConfig := sock5_session.Response{
+		Port: m.servicePort,
+	}
+	sessionDestroy := func() {
+		// Do nothing
+	}
+
+	sessionRequest := sock5_session.Request{}
+	err := json.Unmarshal(sessionRequestData, &sessionRequest)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not unmarshal SOCKS5 session request")
+	}
+
+	natPingerEnabled := m.natPinger.Valid() && m.isBehindNAT() && m.portMappingFailed()
+	traversalParams, err := m.newTraversalParams(natPingerEnabled, sessionRequest)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not create traversal params")
+	}
+
+	if natPingerEnabled {
+		log.Info().Msgf("NAT Pinger enabled, binding service port for proxy, key: %s", traversalParams.ProxyPortMappingKey)
+		m.natPinger.BindServicePort(traversalParams.ProxyPortMappingKey, sessionConfig.Port)
+		sessionConfig.PortNat = traversalParams.ConsumerPort
+		sessionConfig.Port = traversalParams.ProviderPort
+	}
+
+	return &session.ConfigParams{SessionServiceConfig: sessionConfig, SessionDestroyCallback: sessionDestroy, TraversalParams: &traversalParams}, nil
+}
+
+// Serve starts service - does block
+func (m *Manager) Serve(providerID identity.Identity) error {
+	releasePorts := m.natPort(m.servicePort)
+	defer releasePorts()
+
+	conf := &socks5_lib.Config{}
+	server, err := socks5_lib.New(conf)
+	if err != nil {
+		return errors.Wrap(err, "could not initiate SOCKS5 proxy server")
+	}
+
+	log.Info().Msg("SOCKS5 service started successfully now")
+	err = m.runServer(server, fmt.Sprintf(":%d", m.servicePort))
+	if err != nil && err != errServerStopped {
+		return errors.Wrap(err, "SOCKS5 proxy server stopped")
+	}
+
+	return nil
+}
+
+// errServerStopped is returned by the Server's Start() method after a call to Stop().
+var errServerStopped = errors.New("server stopped")
+
+// runServer listens on the TCP network address and then handles incoming connections.
+// runServer always returns a non-nil error.
+// After Stop(), the returned error is errServerStopped.
+func (m *Manager) runServer(server *socks5_lib.Server, address string) error {
+	var err error
+
+	m.listenerLock.Lock()
+	m.listenerStop = make(chan struct{})
+	m.listener, err = net.Listen("tcp", address)
+	m.listenerLock.Unlock()
+	if err != nil {
+		return errors.Wrap(err, "could not initiate SOCKS5 proxy listener")
+	}
+
+	for {
+		conn, err := m.listener.Accept()
+		if err != nil {
+			select {
+			case <-m.listenerStop:
+				return errServerStopped
+			default:
+				return errors.Wrap(err, "input error from SOCKS5 connection")
+			}
+		}
+		go server.ServeConn(conn)
+	}
+}
+
+// Stop stops service.
+func (m *Manager) Stop() error {
+	log.Info().Msg("SOCKS5 service stopping..")
+
+	m.listenerLock.Lock()
+	close(m.listenerStop)
+	m.listenerLock.Unlock()
+
+	if m.listener != nil {
+		return m.listener.Close()
```

Comment:
remove this, and just close the listener if stop is called from the for loop in start method.